core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -1793,16 +1761,51 @@ private <C extends Comparable<C>> RexNode simplifyAnd2ForUnknownAsFalse(
         return rexBuilder.makeLiteral(false);
       }
     }
-    // Add the NOT disjunctions back in.
-    for (RexNode notDisjunction : notTerms) {
-      terms.add(not(notDisjunction));
-    }
     // The negated terms: only deterministic expressions
     for (RexNode negatedTerm : negatedTerms) {
       if (termsSet.contains(negatedTerm)) {
         return rexBuilder.makeLiteral(false);
       }
     }
+    // Add the NOT disjunctions back in.
```

Comment:
Just for my understanding, why you decided to move this part after the for loop above?

Reply:
That's probably does not matter a lot, I have just moved it a bit closer to the code where we actually need this.
I rely on having both `terms` and `notTerms` included to `terms` list during calculation of `strongOperands`.